librosa/core/spectrum.py

```diff
@@ -519,6 +520,450 @@ def ifgram(y, sr=22050, n_fft=2048, hop_length=None, win_length=None,
+
+    sr : number > 0 [scalar]
+        sampling rate of `y`
+
+    n_fft : int > 0 [scalar]
+        FFT window size
+
+    hop_length : int > 0 [scalar]
+        hop length, number samples between subsequent frames.
+        If not supplied, defaults to `win_length / 4`.
+
+    win_length : int > 0, <= n_fft
+        Window length. Defaults to `n_fft`.
+        See `stft` for details.
+
+    window : string, tuple, number, function, or np.ndarray [shape=(n_fft,)]
+        - a window specification (string, tuple, number);
+          see `scipy.signal.get_window`
+        - a window function, such as `scipy.signal.hanning`
+        - a user-specified window vector of length `n_fft`
+
+        See `stft` for details.
+
+        .. see also:: `filters.get_window`
+
+    dtype : numeric type
+        Complex numeric type for `S`. Default is 64-bit complex.
+
+    Returns
+    -------
+    freqs : np.ndarray [shape=(1 + n_fft/2, t), dtype=real]
+        Instantaneous frequencies:
+        `freqs[f, t]` is the frequency for bin `f`, frame `t`
+
+    S : np.ndarray [shape=(1 + n_fft/2, t), dtype=complex]
+        Short-time Fourier transform
+
+    Warns
+    --------
+    RuntimeWarning
+        Frequencies with zero support will produce a divide-by-zero warning and
+        will be returned as `np.nan`.
+
+    See Also
+    --------
+    stft : Short-time Fourier Transform
+    reassign_times : Estimate time reassignments
+    reassigned_spectrogram : Time-frequency reassigned spectrogram
+
+    Examples
+    --------
+    >>> y, sr = librosa.load(librosa.util.example_audio_file())
+    >>> frequencies, S = librosa.reassign_frequencies(y, sr=sr)
+    >>> frequencies
+    array([[  0.000e+00, 0.000e+00, ..., 0.000e+00, 0.000e+00],
+           [  9.863e+00, 9.653e+00, ..., 1.046e+01, 8.072e+01],
+           ...,
+           [  1.101e+04, 1.101e+04, ..., 1.102e+04, 1.102e+04],
+           [  1.102e+04, 1.102e+04, ..., 1.102e+04, 1.102e+04]])
+
+    """
+
+    # retrieve window samples if needed so that the window derivative can be
+    # calculated
+    if win_length is None:
+        win_length = n_fft
+
+    window = get_window(window, win_length, fftbins=True)
+    window = util.pad_center(window, n_fft)
+
+    # center is always False for consistency with `reassign_times`
+    S_h = stft(
+        y=y,
+        n_fft=n_fft,
+        hop_length=hop_length,
+        window=window,
+        center=False,
+        dtype=dtype,
+    )
+
+    # cyclic gradient to correctly handle edges of a periodic window
+    window_derivative = util.cyclic_gradient(window)
+
+    S_dh = stft(
+        y=y,
+        n_fft=n_fft,
+        hop_length=hop_length,
+        window=window_derivative,
+        center=False,
+        dtype=dtype,
+    )
+
+    # equation 5.20 of Flandrin, Auger, & Chassande-Mottin 2002
+    # the sign of the correction is reversed in some papers - see Plante,
+    # Meyer, & Ainsworth 1998 pp. 283-284
+    correction = -np.imag(S_dh / S_h)
```

Comment:
This division makes me nervous.  Is there a more numerically stable way to compute this?